File: DSA/Java/tests/dsa/lib/algo/sort/CountingSortTest.java
package dsa.lib.algo.sort;

import org.junit.Test;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertEquals;

public class CountingSortTest {

    @Test
    public void minMax() {
        assertMinMax(new int[]{1, 1}, 1, 1);
        assertMinMax(new int[]{2, 1}, 1, 2);
        assertMinMax(new int[]{1, 2, 3}, 1, 3);
        assertMinMax(new int[]{1, 1, 1}, 1, 1);
        assertMinMax(new int[]{4, 2, 8}, 2, 8);
    }

    private void assertMinMax(int[] nums, int min, int max) {
        CountingSort cs = getCountingSort(nums);
        assertEquals(min, cs.min);
        assertEquals(max, cs.max);
    }

    @Test
    public void countsSize() {
        assertCountsSize(new int[]{0, 0}, 1);
        assertCountsSize(new int[]{1, 1}, 1);
        assertCountsSize(new int[]{0, 1, 1}, 2);
        assertCountsSize(new int[]{2, 4, 8}, 7);
    }

    private void assertCountsSize(int[] nums, int expected) {
        assertEquals(expected, getCountingSort(nums).counts.length);
    }

    @Test
    public void countsIndexOf() {
        assertCountsIndexOf(new int[]{1, 2, 3}, 2, 1);
        assertCountsIndexOf(new int[]{1, 4, 1}, 4, 3);
        assertCountsIndexOf(new int[]{3, 4, 2}, 2, 0);
    }

    private void assertCountsIndexOf(int[] nums, int num, int expectedIndex) {
        assertEquals(expectedIndex, getCountingSort(nums).countsIndexOf(num));
    }

    private CountingSort getCountingSort(int[] nums) {
        CountingSort cs = new CountingSort(nums);
        cs.sort();
        return cs;
    }

    @Test
    public void emptyArray() {
        assertCountingSort(new int[]{}, new int[]{});
    }

    @Test
    public void oneElement() {
        assertCountingSort(new int[]{1}, new int[]{1});
    }

    @Test
    public void sorts() {
        assertCountingSort(new int[]{1, 1, 0}, new int[]{0, 1, 1});
        assertCountingSort(new int[]{3, 2, 1}, new int[]{1, 2, 3});
        assertCountingSort(new int[]{4, 5, 6, 3, 2, 1}, new int[]{1, 2, 3, 4, 5, 6});
        assertCountingSort(new int[]{12, 11, 13, 5, 6, 7}, new int[]{5, 6, 7, 11, 12, 13});
        assertCountingSort(new int[]{2, 5, 3, 0, 2, 3, 0, 3}, new int[]{0, 0, 2, 2, 3, 3, 3, 5});
    }

    private void assertCountingSort(int[] nums, int[] expected) {
        countingSort(nums);
        assertArrayEquals(expected, nums);
    }

    private static void countingSort(int[] ary) {
        new CountingSort(ary).sort();
    }
}

File: DSA/Cpp/test/dsa/lib/algo/sort/counting_sort_test.cpp
#include <dsa/lib/algo/sort/counting_sort.hpp>
#include "sort_test.h"

namespace dsa::lib::algo::sort {
namespace {

class CountingSortTest : public SortTest {
 protected:
    static void assertCountingSort(std::vector<int> &&expected, std::vector<int> &&actual) {
        countingSort(actual);
        EXPECT_EQ(expected, actual);
    }

    static void countingSort(std::vector<int> &nums) {
        CountingSort(nums).sort();
    }

    static void assertMinMax(std::vector<int> &&nums, int min, int max) {
        CountingSort cs(nums);
        cs.sort();
        EXPECT_EQ(min, cs.min);
        EXPECT_EQ(max, cs.max);
    }

    static void assertCountsSize(std::vector<int> &&nums, size_t size) {
        CountingSort cs(nums);
        cs.sort();
        EXPECT_EQ(size, cs.counts->size());
    }

    static void assertCountsIndexOf(std::vector<int> &&nums, int num, int index) {
        CountingSort cs(nums);
        cs.sort();
        EXPECT_EQ(index, cs.countsIndexOf(num));
    }
};
TEST_F(CountingSortTest, minMax) {
    assertMinMax({1, 1}, 1, 1);
    assertMinMax({1, 2, 3}, 1, 3);
    assertMinMax({8, 2, 9}, 2, 9);
}
TEST_F(CountingSortTest, countsSize) {
    assertCountsSize({1, 1}, 1);
    assertCountsSize({1, 2, 3}, 3);
    assertCountsSize({8, 2, 9}, 8);
}
TEST_F(CountingSortTest, countsIndexOf) {
    assertCountsIndexOf({1, 1}, 1, 0);
    assertCountsIndexOf({1, 2, 3}, 2, 1);
    assertCountsIndexOf({8, 2, 9}, 9, 7);
}
TEST_F(CountingSortTest, emptyArray) {
    assertCountingSort({}, {});
}
TEST_F(CountingSortTest, oneElement) {
    assertCountingSort({1}, {1});
}
TEST_F(CountingSortTest, sorts) {
    assertCountingSort({0, 1, 1}, {1, 1, 0});
    assertCountingSort({1, 2, 3}, {3, 2, 1});
    assertCountingSort({1, 2, 3, 4, 5, 6}, {4, 5, 6, 3, 2, 1});
    assertCountingSort({5, 6, 7, 11, 12, 13}, {12, 11, 13, 5, 6, 7});
    assertCountingSort({0, 0, 2, 2, 3, 3, 3, 5}, {2, 5, 3, 0, 2, 3, 0, 3});
}

}
}
